<a href="https://colab.research.google.com/github/FauziArifin/PacktData/blob/master/LetterName.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Train a Random Forest Classifier on the ISOLET Dataset**


---
You are working for a technology company and they are planning to launch a new voice assistant product. You have been tasked with building a classification model that will recognize the letters spelled out by a user based on the signal frequencies captured. Each sound can be captured and represented as a signal composed of multiple frequencies.


**Just Predict with Random Forest Classifier**

In [53]:
import pandas as pd

url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter04/Dataset/phpB0xrNj.csv'

df = pd.read_csv(url)

df.head()

f1      f2      f3      f4  ...    f615    f616    f617  class
0 -0.4394 -0.0930  0.1718  0.4620  ...  0.6410  0.5898 -0.4872    '1'
1 -0.4348 -0.1198  0.2474  0.4036  ...  0.4318  0.4546 -0.0910    '1'
2 -0.2330  0.2124  0.5014  0.5222  ...  0.2540  0.1588 -0.4762    '2'
3 -0.3808 -0.0096  0.2602  0.2554  ... -0.0072  0.0936 -0.1510    '2'
4 -0.3412  0.0946  0.6082  0.6216  ...  0.3124  0.2188 -0.2500    '3'

[5 rows x 618 columns]

In [54]:
df.dtypes

f1       float64
f2       float64
f3       float64
f4       float64
f5       float64
          ...   
f614     float64
f615     float64
f616     float64
f617     float64
class     object
Length: 618, dtype: object

In [55]:
df.describe().T

count      mean       std     min     25%     50%     75%     max
f1    7797.0 -0.385396  0.236721 -1.0000 -0.5468 -0.4140 -0.2410  1.0000
f2    7797.0  0.141103  0.320090 -0.8926 -0.0994  0.1064  0.3546  1.0000
f3    7797.0  0.347384  0.326658 -0.9752  0.1054  0.3334  0.5834  1.0000
f4    7797.0  0.450659  0.308301 -0.9680  0.2478  0.4532  0.6686  1.0000
f5    7797.0  0.324283  0.461491 -0.9966 -0.0248  0.4158  0.6924  1.0000
...      ...       ...       ...     ...     ...     ...     ...     ...
f613  7797.0  0.299238  0.334282 -0.7708  0.0492  0.2800  0.5412  1.0000
f614  7797.0  0.224122  0.336135 -0.8106 -0.0178  0.1924  0.4616  1.0000
f615  7797.0  0.186668  0.337780 -0.7802 -0.0576  0.1578  0.4108  1.0000
f616  7797.0  0.081370  0.332859 -0.8028 -0.1546  0.0452  0.2898  1.0000
f617  7797.0 -0.281416  0.354418 -0.9626 -0.5466 -0.3290 -0.0656  0.8898

[617 rows x 8 columns]

In [56]:
df.isnull().sum()

f1       0
f2       0
f3       0
f4       0
f5       0
        ..
f614     0
f615     0
f616     0
f617     0
class    0
Length: 618, dtype: int64

In [57]:
df.count(0)

f1       7797
f2       7797
f3       7797
f4       7797
f5       7797
         ... 
f614     7797
f615     7797
f616     7797
f617     7797
class    7797
Length: 618, dtype: int64

In [58]:
df['class'].unique()

array(["'1'", "'2'", "'3'", "'4'", "'5'", "'6'", "'7'", "'8'", "'9'",
       "'10'", "'11'", "'12'", "'13'", "'14'", "'15'", "'16'", "'17'",
       "'18'", "'19'", "'20'", "'21'", "'22'", "'23'", "'24'", "'25'",
       "'26'"], dtype=object)

In [0]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.feature_extraction.text import TfidfTransformer

X = df.drop(['class'], axis = 1)
Y = df['class']

# scaler = MinMaxScaler()
# X = scaler.fit_transform(X)

tfidf = TfidfTransformer()
X = tfidf.fit_transform(X)

In [64]:
print(X.shape)

(7797, 617)


In [66]:
Y.shape

(7797,)

In [0]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state=888)

In [0]:
from sklearn.ensemble import RandomForestClassifier

#Create a function called train_rf
def train_rf(x_train, y_train, random_state=888, n_estimators=10, max_depth=None, min_samples_leaf=1, max_features='sqrt'):
  rf_model = RandomForestClassifier(random_state=random_state, n_estimators=n_estimators, max_depth=max_depth, min_samples_leaf=min_samples_leaf, max_features=max_features)
  rf_model.fit(x_train, y_train)
  return rf_model

In [0]:
#Create a function called get_preds that will take as arguments rf_model, x_train, and x_test
def get_preds(rf_model, x_train, x_test):
  train_preds = rf_model.predict(x_train)
  test_preds = rf_model.predict(x_test)
  return train_preds, test_preds

In [0]:
from sklearn.metrics import accuracy_score

#Create a function called print_accuracy
def print_accuracy(y_train, y_test, train_preds, test_preds):
  accuracy_train = accuracy_score(y_train, train_preds)
  accuracy_test = accuracy_score(y_test, test_preds)
  print('Accuracy score for train dataset : ', accuracy_train)
  print('Accuracy score for test dataset : ', accuracy_test)
  return accuracy_train, accuracy_test

In [71]:
#fit Random forest using train dataset
rf_model1 = train_rf(x_train, y_train)

#predict
predict_train, predict_test = get_preds(rf_model1, x_train, x_test)

#show accuracy score
predict_train, predict_test = print_accuracy(y_train, y_test, predict_train, predict_test)

Accuracy score for train dataset :  0.9979156645823313
Accuracy score for test dataset :  0.8955128205128206


In [0]:
#Create a function fit_predict_RF
def fit_predict_RF(x_train, x_test, y_train, y_test, random_state=888, n_estimators=10, max_depth=None, min_samples_leaf=1, max_features='sqrt'):
  rf_model = train_rf(x_train, y_train, random_state=random_state, n_estimators=n_estimators, max_depth=max_depth, min_samples_leaf=min_samples_leaf, max_features=max_features)
  train_preds, test_preds = get_preds(rf_model, x_train, x_test)
  accuracy_train, accuracy_test = print_accuracy(y_train, y_test, train_preds, test_preds)
  return rf_model, train_preds, test_preds, accuracy_train, accuracy_test
  

In [83]:
#Try another parameter for Random Forest

#with n_estimators=20
rf_model2, train_preds2, test_preds2, accuracy_train2, accuracy_test2 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=20)

Accuracy score for train dataset :  0.9996793330126663
Accuracy score for test dataset :  0.9237179487179488


Compared to the previous results, our model is overfitting less. This is increase the accuracy score of the testing set and therefore decreased the difference with the training set

In [86]:
#with n_estimators=50
rf_model3, train_preds3, test_preds3, accuracy_train3, accuracy_test3 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50)

Accuracy score for train dataset :  1.0
Accuracy score for test dataset :  0.9365384615384615


With this set of hyperparameters, obtaine a better accuracy score for the training and testing sets but the model is still overfitting.

In [87]:
#with n_estimators=50, max_depth=5
rf_model4, train_preds4, test_preds4, accuracy_train4, accuracy_test4 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50, max_depth=5)

Accuracy score for train dataset :  0.8682058682058682
Accuracy score for test dataset :  0.823076923076923


Now the model performance has decreased drastically but the model is not overfitting much

In [90]:
#with different max_depth : max_depth=10
rf_model5, train_preds5, test_preds5, accuracy_train5, accuracy_test5 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50, max_depth=10)

Accuracy score for train dataset :  0.9838063171396505
Accuracy score for test dataset :  0.9237179487179488


The accuracy scores for the training and testing sets are quite high but their difference is still significant.

In [91]:
#with different min_samples_leaf : min_samples_leaf=10
rf_model6, train_preds6, test_preds6, accuracy_train6, accuracy_test6 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50, max_depth=10, min_samples_leaf=10)

Accuracy score for train dataset :  0.9631232964566298
Accuracy score for test dataset :  0.9147435897435897


With this set of hyperparameters, start to reach a good compromise between performance and overfitting

In [92]:
#with different min_samples_leaf : min_samples_leaf=50
rf_model7, train_preds7, test_preds7, accuracy_train7, accuracy_test7 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50, max_depth=10, min_samples_leaf=50)

Accuracy score for train dataset :  0.9259259259259259
Accuracy score for test dataset :  0.9076923076923077


Now the accuracy scores for the training and testing sets are quite close to each other and we still have a good level of performance (around 0.9).

In [93]:
#with different max_features : max_features=0.5
rf_model8, train_preds8, test_preds8, accuracy_train8, accuracy_test8 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50, max_depth=10, min_samples_leaf=50, max_features=0.5)

Accuracy score for train dataset :  0.8941798941798942
Accuracy score for test dataset :  0.8743589743589744


The model is not achieving the same level as for the previous result.

In [94]:
#with different max_features : max_features=0.2
rf_model8, train_preds8, test_preds8, accuracy_train8, accuracy_test8 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=50, max_depth=10, min_samples_leaf=50, max_features=0.2)

Accuracy score for train dataset :  0.9076479076479076
Accuracy score for test dataset :  0.8858974358974359


still doesn't achieve better results

In [0]:
# #with different min_samples_leaf : n_estimators=100
# rf_model9, train_preds9, test_preds9, accuracy_train9, accuracy_test9 = fit_predict_RF(x_train, x_test, y_train, y_test,n_estimators=100, max_depth=10, min_samples_leaf=50)

Final accuracy score of 0.92 for the training set and 0.90 for the testing set. The model is still overfitting slightly and could still be improved but it is a remarkable result given that it was our first attempt.